crates/matrix-sdk-base/src/store/mod.rs

```diff
@@ -319,6 +270,175 @@ impl Deref for Store {
+/// we have a redirection from a key to a position, and from a position to
+/// a value. The (key, position) tuples are stored in an [`HashMap`]. The
+/// (position, value) tuples are stored in an [`ObservableVector`]. The (key,
+/// position) tuple is only provided for fast _reading_ implementations, like
+/// `Self::get` and `Self::get_or_create`. The (position, value) tuples are
+/// observable, this is what interests us the most here.
+///
+/// Why not implementing a new `ObservableMap` type in `eyeball-im` instead of
+/// this custom implementation? Because we want to continue providing
+/// `VectorDiff` when observing the changes, so that the rest of the API in the
+/// Matrix Rust SDK aren't broken. Indeed, an `ObservableMap` must produce
+/// `MapDiff`, which would be quite different.
+/// Plus, we would like to re-use all our existing code, test, stream adapters
+/// and so on.
+///
+/// This is a trade-off. And this implementation is simple enough for the
+/// moment, and basically does the job.
+#[derive(Debug)]
+struct ObservableMap<K, V>
+where
+    V: Clone + Send + Sync + 'static,
+{
+    /// The (key, position) tuples.
+    mapping: HashMap<K, usize>,
+
+    /// The values where the indices are the `position` part of `Self::mapping`.
+    values: ObservableVector<V>,
+}
+
+impl<K, V> ObservableMap<K, V>
+where
+    K: Hash + Eq,
+    V: Clone + Send + Sync + 'static,
+{
+    /// Create a new `Self`.
+    fn new() -> Self {
+        Self { mapping: HashMap::new(), values: ObservableVector::new() }
+    }
+
+    /// Insert a new `V` in the collection.
+    ///
+    /// The position of the inserted value is returned.
+    ///
+    /// If the `V` value already exists, it will be updated to the new one.
+    fn insert(&mut self, key: K, value: V) -> usize {
+        match self.mapping.get(&key) {
+            Some(position) => {
+                self.values.set(*position, value);
+
+                *position
+            }
+            None => {
+                let position = self.values.len();
+
+                self.values.push_back(value);
+                self.mapping.insert(key, position);
+
+                position
+            }
+        }
+    }
+
+    /// Reading one `V` value based on their ID, if it exists.
+    fn get<L>(&self, key: &L) -> Option<&V>
+    where
+        K: Borrow<L>,
+        L: Hash + Eq + ?Sized,
+    {
+        self.mapping.get(key).and_then(|position| self.values.get(*position))
+    }
+
+    /// Reading one `V` value based on their ID, or create a new one (by using
+    /// `default`).
+    fn get_or_create<L, F>(&mut self, key: &L, default: F) -> &V
+    where
+        K: Borrow<L>,
+        L: Hash + Eq + ?Sized + ToOwned<Owned = K>,
+        F: FnOnce() -> V,
+    {
+        let position = match self.mapping.get(key) {
+            Some(position) => *position,
+            None => {
+                let value = default();
+                self.insert(key.to_owned(), value)
```

Comment:
(Same comment as previous applies here too, we may be doing the lookup once too many)